engine/grammar/matchers/datatypes/semver.go

```diff
@@ -0,0 +1,202 @@
+package datatypes
+
+import (
+	"fmt"
+	"strconv"
+
+	"strings"
+)
+
+const (
+	MetadataDelimiter   = "+"
+	PreReleaseDelimiter = "-"
+	ValueDelimiter      = "."
+)
+
+type Semver struct {
+	major     int64
+	minor     int64
+	patch     int64
+	preRealse []string
+	isStable  bool
+	metadata  string
+	version   string
+}
+
+func BuildSemver(version string) (*Semver, error) {
+	if len(strings.TrimSpace(version)) == 0 {
+		return nil, fmt.Errorf("unable to convert to semver, version cannot be empty")
+	}
+	metadata, vWithoutMetadata, err := processMetadata(version)
+	if err != nil {
+		return nil, err
+	}
+	preRealse, vWithoutPreRelease, err := processPreRelease(vWithoutMetadata)
+	if err != nil {
+		return nil, err
+	}
+	major, minor, patch, err := processComponents(vWithoutPreRelease)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Semver{
+		metadata:  metadata,
+		preRealse: preRealse,
+		major:     major,
+		minor:     minor,
+		patch:     patch,
+		isStable:  len(preRealse) == 0,
+		version:   version,
+	}, nil
+}
+
+func processMetadata(version string) (string, string, error) {
+	index, metadata := calculateIndexAndTrimString(version, MetadataDelimiter)
+	if index == -1 {
+		return "", version, nil
```

Comment:
probably missing TrimSpace(version)

Reply:
It's not necessary. When we process components, we'll return an error if it can't be parsed to int.